### PR TITLE
new: asking for review notice on admin dashboard pages

### PR DESCRIPTION
**Pull request by devAsadNur**
Added asking for review notice on admin dashboard pages.
Screenshot: https://imgur.com/a/C3MHWQI

### CODE REVIEW

**Comment by shanjidah** [comment]
Scenario: There is one warning shows on the dashboard.

**Here is the screenshot:**
Screen Shot 2021-11-05 at 12 03 10 PM

File: includes/Admin/ReviewNotice.php [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+        if ( in_array( $pagenow, $exclude, true ) ) {
+            return;
+        }
+
+        if ( ! current_user_can( 'manage_woocommerce' ) ) {
+            return;
+        }
+
+        if ( empty( get_option( 'dokan_review_notice_enable', 1 ) ) ) {
```
Comment (nurul-umbhiya) [inline includes/Admin/ReviewNotice.php:46]: dokan_review_notiice_hidden

File: includes/Admin/ReviewNotice.php [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        $updated = [];
+        if ( 'dokan-notice-dismiss' === wp_unslash( sanitize_key( $_POST['key'] ) ) ) {
+            $updated[] = update_option( 'dokan_review_notice_enable', 0 ) ? 1 : 0;
+
+            delete_option( 'dokan_review_notice_postpond' );
+            delete_option( 'dokan_review_notice_postpond_time' );
+        }
+
+        if ( 'dokan-notice-postpond' === wp_unslash( sanitize_key( $_POST['key'] ) ) ) {
+            $updated[] = update_option( 'dokan_review_notice_enable', 1 ) ? 1 : 0;
```
Comment (nurul-umbhiya) [inline includes/Admin/ReviewNotice.php:97]:
```
$postponed_days = apply_filters( 'dokan_ask_for_review_admin_notice_postpond_days', 15 );
$postponed = dokan_current_datetime()->modify( "+$postponed_dayes days" )->getTimestamp();
update_option( 'dokan_review_notice_postponed', $postponed );
wp_send_json_success();
```

File: includes/Admin/ReviewNotice.php [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+        if ( empty( $_POST['dokan_ask_for_review_notice_action'] ) || empty( $_POST['key'] ) ) {
+            wp_send_json_error( __( 'Invalid request', 'dokan-lite' ) );
+        }
+
+        $updated = [];
+        if ( 'dokan-notice-dismiss' === wp_unslash( sanitize_key( $_POST['key'] ) ) ) {
+            $updated[] = update_option( 'dokan_review_notice_enable', 0 ) ? 1 : 0;
+
+            delete_option( 'dokan_review_notice_postpond' );
```
Comment (nurul-umbhiya) [inline includes/Admin/ReviewNotice.php:92]: remove dokan_review_notice_postponed

File: includes/Admin/ReviewNotice.php (hunk 1 above)
Comment (nurul-umbhiya) [inline includes/Admin/ReviewNotice.php:46]: replace below code with:

```
// check if notice was hidden
if ( 'yes' === get_option( 'dokan_review_notice_hidden', 'no' ) ) {
    return;
}

// check if notice postponed
$notice_postponed = get_option( 'dokan_review_notice_postponed' );
if ( ! empty( $notice_postponed )  && $notice_postponed < time() ) {
    return;
}

// check if install time exist
$installed_time = get_option( 'dokan_installed_time' );
if ( ! empty( $installed_time ) ) {
    $initial_delay_days = apply_filters( 'dokan_ask_for_review_admin_notice_initial_delay_days', 10 );
    $eligible_time = dokan_current_datetime()->modify( "+$initial_delay_days  days" );
    if ( $eligible_time->getTimestamp() < time() ) {
        return;
    }
}

// all checking passed, display  admin notice
dokan_get_template( 'admin-notice/ask-for-review.php' );
```

File: includes/Install/Installer.php [hunk 4]
```diff
@@ -45,6 +46,22 @@ public function do_install() {
...
 
+    /**
+     * Adds plugin installation time.
+     *
+     * @since DOKAN_LITE_SINCE
+     *
+     * @return boolean
+     */
+    public function add_version_info() {
+        $first_installed = get_option( 'dokan_installed_time' );
```
Comment (nurul-umbhiya) [inline includes/Install/Installer.php:57]:
```
if ( empty( get_option( 'dokan_installed_time' ) ) ) {
    $current_time  = dokan_current_datetime()->getTimestamp();
    update_option( 'dokan_installed_time', $current_time );
}
```

File: assets/src/js/dokan-admin-notice.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
+;(function($) {
```
Comment (nurul-umbhiya) [inline assets/src/js/dokan-admin-notice.js:1]: move this code block to template file

File: assets/src/js/dokan-admin-notice.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        init : function () {
+            $( '.dokan-review-notice .dokan-notice-action' ).on( 'click', function( e ) {
+                // Prevent default action if the target is not a link
+                if ( 'a' !== e.target.tagName.toLowerCase() ) {
+                    e.preventDefault();
+                }
+
+                let self  = $( this );
+                let key   = self.data( 'key' );
+                let nonce = self.data( 'nonce' );
```
Comment (nurul-umbhiya) [inline assets/src/js/dokan-admin-notice.js:16]: remove  this line

File: assets/src/js/dokan-admin-notice.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+                wp.ajax.send( 'ask_for_review_admin_notice', {
+                    data: {
+                        dokan_ask_for_review_notice_action: true,
+                        key: key,
+                        nonce: nonce,
```
Comment (nurul-umbhiya) [inline assets/src/js/dokan-admin-notice.js:22]:
```
nonce: '<?php echo wp_create_nonce( 'dokan_admin_review_notice_nonce' ); ?>'
```

File: includes/Admin/ReviewNotice.php [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+    /**
+     * Reveiw notice action ajax handler
+     *
+     * @since 3.2.16
+     *
+     * @return void
+     */
+    public function review_notice_action_handler() {
+        if ( ! isset( $_POST['nonce'] ) || ! wp_verify_nonce( sanitize_key( wp_unslash( $_POST['nonce'] ) ), 'dokan_admin' ) ) {
```
Comment (nurul-umbhiya) [inline includes/Admin/ReviewNotice.php:76]: change `dokan_admin` action  with  `dokan_admin_review_notice_nonce`

File: assets/src/js/dokan-admin-notice.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+                    e.preventDefault();
+                }
+
+                let self  = $( this );
+                let key   = self.data( 'key' );
+                let nonce = self.data( 'nonce' );
+
+                wp.ajax.send( 'ask_for_review_admin_notice', {
+                    data: {
+                        dokan_ask_for_review_notice_action: true,
```
Comment (nurul-umbhiya) [inline assets/src/js/dokan-admin-notice.js:20]: remove this line

File: includes/Admin/ReviewNotice.php [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+            wp_send_json_error( __( 'Invalid nonce', 'dokan-lite' ) );
+        }
+
+        if ( ! current_user_can( 'manage_woocommerce' ) ) {
+            wp_send_json_error( __( 'You have no permission to do that', 'dokan-lite' ) );
+        }
+
+        if ( empty( $_POST['dokan_ask_for_review_notice_action'] ) || empty( $_POST['key'] ) ) {
```
Comment (nurul-umbhiya) [inline includes/Admin/ReviewNotice.php:84]: change this line with:
```
$key = isset( $_POST['key'] ) ? sanitize_text_fields( wp_unslash( $_POST['key'] ) ) : '';
if ( empty( $key ) || ! in_array( $key, [ 'dokan-notice-postpond', 'dokan-notice-dismiss' ],  true ) ) {
    wp_send_json_error( __( 'Invalid request', 'dokan-lite' ) );
}
```
```

File: includes/Admin/ReviewNotice.php [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+            wp_send_json_error( __( 'Invalid request', 'dokan-lite' ) );
+        }
+
+        $updated = [];
+        if ( 'dokan-notice-dismiss' === wp_unslash( sanitize_key( $_POST['key'] ) ) ) {
```
Comment (nurul-umbhiya) [inline includes/Admin/ReviewNotice.php:89]: 'dokan-notice-dismiss' === $key

File: includes/Admin/ReviewNotice.php (hunk 3 above)
Comment (nurul-umbhiya) [reply includes/Admin/ReviewNotice.php:92]: wp_send_json_success();

**Review by nurul-umbhiya — APPROVED** [review]

**Review by nurul-umbhiya — APPROVED** [review]

**Pull request merged**